habitat_baselines/train_ppo.py

```diff
@@ -157,13 +165,127 @@ def construct_envs(args):
     return envs
 
 
+def rollout_step(
+    args,
+    envs,
+    agent,
+    rollouts,
+    current_episode_reward,
+    episode_rewards,
+    episode_counts,
+):
+
+    pth_time = 0
+    env_time = 0
+    count_steps = 0
+    actor_critic = agent.actor_critic
+
+    t_sample_action = time()
+    # sample actions
+    with torch.no_grad():
+        step_observation = {
+            k: v[rollouts.step] for k, v in rollouts.observations.items()
+        }
+
+        (
+            values,
+            actions,
+            actions_log_probs,
+            recurrent_hidden_states,
+        ) = actor_critic.act(
+            step_observation,
+            rollouts.recurrent_hidden_states[rollouts.step],
+            rollouts.prev_actions[rollouts.step],
+            rollouts.masks[rollouts.step],
+        )
+
+    pth_time += time() - t_sample_action
+
+    t_step_env = time()
+
+    outputs = envs.step([a[0].item() for a in actions])
+    observations, rewards, dones, infos = [list(x) for x in zip(*outputs)]
+
+    env_time += time() - t_step_env
+
+    t_update_stats = time()
+    batch = batch_obs(observations)
+    rewards = torch.tensor(rewards, dtype=torch.float)
+    rewards = rewards.unsqueeze(1)
+
+    masks = torch.tensor(
+        [[0.0] if done else [1.0] for done in dones], dtype=torch.float
+    )
+
+    device = current_episode_reward.device
+    rewards = rewards.to(device)
+    masks = masks.to(device)
+    current_episode_reward += rewards
+    episode_rewards += (1 - masks) * current_episode_reward
+    episode_counts += 1 - masks
+    current_episode_reward *= masks
+
+    rollouts.insert(
+        batch,
+        recurrent_hidden_states,
+        actions,
+        actions_log_probs,
+        values,
+        rewards,
+        masks,
+    )
+
+    count_steps += envs.num_envs
+    pth_time += time() - t_update_stats
+
+    return pth_time, env_time, count_steps
+
+
+def update_agent(args, agent, rollouts):
+    actor_critic = agent.actor_critic
+    actor_critic.train()
+
+    t_update_model = time()
+    with torch.no_grad():
+        last_observation = {k: v[-1] for k, v in rollouts.observations.items()}
+        next_value = actor_critic.get_value(
+            last_observation,
+            rollouts.recurrent_hidden_states[-1],
+            rollouts.prev_actions[-1],
+            rollouts.masks[-1],
+        ).detach()
+
+    rollouts.compute_returns(next_value, args.use_gae, args.gamma, args.tau)
+
+    value_loss, action_loss, dist_entropy = agent.update(rollouts)
+
+    rollouts.after_update()
+    return time() - t_update_model, value_loss, action_loss, dist_entropy
+
+
 def run_training():
     parser = ppo_args()
+    parser.add_argument(
```

Comment:
Why not to add to `ppo_args` ?

Reply:
The `sim-gpu-id` and `pth-gpu-id` are specific to this train script.  Different algos may have different ways to figure out which GPU to use (i.e. DD-PPO).